src/libxively/event_dispatcher/xi_time_event.h

```diff
@@ -0,0 +1,128 @@
+/* Copyright (c) 2003-2015, LogMeIn, Inc. All rights reserved.
+ * This is part of Xively C library. */
+
+/**
+ * @file xi_time_event.h
+ * @brief Implements time based events
+ *
+ * Time based events can be used to register function handler for later execution.
+ * Entities such timeouts can be easily implemented using time based events.
+ */
+
+#ifndef __XI_TIME_EVENT_H__
+#define __XI_TIME_EVENT_H__
+
+#include <stdint.h>
+
+#include "xi_vector.h"
+#include "xi_debug.h"
+#include "xi_allocator.h"
+#include "xi_time.h"
+#include "xi_event_handle.h"
+
+typedef struct xi_time_event_handle_s
+{
+    xi_vector_index_type_t* position;
+} xi_time_event_handle_t;
+
+typedef struct xi_time_event_s
+{
+    xi_event_handle_t event_handle;
+    xi_time_t time_of_execution;
+    xi_vector_index_type_t position;
+    xi_time_event_handle_t* time_event_handle;
+} xi_time_event_t;
+
+#define XI_TIME_EVENT_POSITION_INVALID -1
+
+#define xi_make_empty_time_event_handle()                                                \
+    {                                                                                    \
+        NULL                                                                             \
+    }
+
+#define xi_make_time_event_handle( event_handle )                                        \
+    {                                                                                    \
+        &event_handle->position                                                          \
+    }
+
+#define xi_make_empty_time_event()                                                       \
+    {                                                                                    \
+        xi_make_empty_event_handle(), 0, XI_TIME_EVENT_POSITION_INVALID, NULL            \
+    }
+
+
+/* API */
+/**
+ * @brief xi_time_event_add
+ *
+ * Adds new time_event to the given vector. If the operation succeded it have to return
```

Comment:
If the operation succeeds, it will return the new time via the parameter ret_time_handle.  The handle will be associated with the provided time_event.  You can use this handle to cancel or restart the time_event (via a call to..? )